test/auth.test.ts

```diff
@@ -0,0 +1,356 @@
+import Auth from '../src/auth';
+import Messenger from '../src/messenger';
+import { EmitterMessage, WorkerMessage } from '../src/types';
+
+describe('Auth', () => {
+  let windowOpenSpy: any;
+
+  beforeEach(() => {
+    windowOpenSpy = jest.spyOn(window, 'open');
+    windowOpenSpy.mockImplementation(() => {});
+    SharedWorker = jest.fn(
+      () =>
+        ({
+          port: {
+            start: () => {},
+            postMessage: () => {},
+          },
+        } as any)
+    );
+  });
+
+  afterEach(() => {
+    windowOpenSpy.mockRestore();
+    jest.clearAllMocks();
+  });
+
+  describe('constructor', () => {
+    it('sets the auth token if passed', () => {
+      const authToken = 'asdf';
+      const clientId = '1234';
+      const storageSpy = jest.spyOn(Storage.prototype, 'setItem');
+
+      new Auth({
+        clientId,
+        host: 'http://localhost:1738',
+        authToken,
+        messenger: new Messenger({}),
+      });
+      expect(storageSpy).toHaveBeenCalledWith('wally:1234:token', authToken);
+
+      storageSpy.mockRestore();
+    });
+  });
+
+  describe('login', () => {
+    const clientId = '1234';
+    const host = 'http://localhost:1738';
+    let auth: Auth;
+
+    beforeEach(() => {
+      auth = new Auth({
+        clientId,
+        host,
+        authToken: 'asdf',
+        messenger: new Messenger({
+          sharedWorkerUrl: '/worker.js',
+        }),
+      });
+    });
+
+    afterEach(() => {
+      auth = undefined as any;
+    });
+
+    it('rejects if already logging in', () => {
+      auth.isLoggingIn = true;
+      return expect(auth.login()).rejects.toBeDefined();
+    });
+
+    it('shows console error if no client id set', () => {
+      const consoleSpy = jest.spyOn(console, 'error');
+      consoleSpy.mockImplementation(() => {});
+      const codeGenSpy = jest.spyOn(Auth.prototype as any, 'generateStateCode');
+      auth = new Auth({} as any);
+      auth.login();
+      expect(consoleSpy).toHaveBeenCalled();
+      expect(codeGenSpy).toHaveBeenCalledTimes(0);
+
+      consoleSpy.mockRestore();
+      codeGenSpy.mockRestore();
```

Comment:
Minor and non-blocking: Similar comment as the above. Could improve the readability a bit to move the spy setup and teardown into the `beforeEach` and `afterEach`.